### PR TITLE
Custom http server example

### PR DESCRIPTION
**Pull request by drisspg**
Hey

### CODE REVIEW

File: toy-submission/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+COPY /lit-llama/ /submission/
+
+# Setup server requriements
+COPY ./fast_api_requirements.txt fast_api_requirements.txt
+RUN pip install --no-cache-dir --upgrade -r fast_api_requirements.txt
+
+# Install any needed packages specified in requirements.txt
+RUN apt-get update && apt-get install -y git
+RUN pip install torch numpy scipy
+RUN pip install -r requirements.txt
```
Comment (msaroufim) [inline toy-submission/Dockerfile:18]: I like the idea of keeping a requirements.txt but this file currently doesn't exist, could you please just move the line above it to a requirements.txt for the torch, numpy and scipy dependencies

File: toy-submission/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+# get open-llama weights
+RUN python scripts/download.py --repo_id openlm-research/open_llama_7b --local_dir checkpoints/open-llama/7B
```
Comment (msaroufim) [inline toy-submission/Dockerfile:21]: Where are the scripts?

File: toy-submission/fast_api_requirements.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# FAST API
```
Comment (msaroufim) [inline toy-submission/fast_api_requirements.txt:1]: add some comment to make it clear that is a suggestion, that people can use whatever server they want

File: toy-submission/helper.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_llama import LLaMA, Tokenizer
+from lit_llama.utils import lazy_load, llama_model_lookup, quantization
+
+@torch.no_grad()
+def neurips_generate(
```
Comment (msaroufim) [inline toy-submission/helper.py]: nit: rename to sample_submission to make it clear that this is not an interface

File: toy-submission/helper.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    model: LLaMA,
+    idx: torch.Tensor,
+    max_new_tokens: int,
+    *,
+    max_seq_length: Optional[int] = None,
+    temperature: float = 1.0,
+    top_k: Optional[int] = None,
+    eos_id: Optional[int] = None,
+) -> Tuple[List[int], List[float], List[Tuple[int, float]]]:
```
Comment (msaroufim) [inline toy-submission/helper.py:24]: can you make this a class so we can put the contract at the very top of this file

File: toy-submission/main.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+model = fabric.setup(model)
+
+tokenizer = Tokenizer(tokenizer_path)
+
+
+class ProcessRequest(BaseModel):
+    prompt: str
+    num_samples: int = 1
+    max_new_tokens: int = 50
+    top_k: int = 200
```
Comment (msaroufim) [inline toy-submission/main.py]: do we need top_k, temperature and seed? As in how do we know for sure if people did use them in their code. I guess if they dont and their results are not reproducible then shucks too bad?

File: toy-submission/main.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    tokens = [Token(text=tokenizer.decode(t), logprob=lp, top_logprob=dict([tlp])) for t, lp, tlp in zip(tokens, logprobs, top_logprobs)]
+    logprobs_sum = sum(logprobs)
+    # Process the input data here
+    return ProcessResponse(text=output, tokens=tokens, logprob=logprobs_sum, request_time=t)
+
+
+class TokenizeRequest(BaseModel):
+        text: str
+        truncation: bool = True
+        max_length: int = 2048
```
Comment (msaroufim) [inline toy-submission/main.py]: should we instead use the state of the art max like 32k?

File: toy-submission/main.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+async def tokenize(input_data: TokenizeRequest) -> TokenizeResponse:
+    logger.info("Using device: {}".format(fabric.device))
+    t0 = time.perf_counter()
+    encoded = tokenizer.encode(input_data.text, bos=True, eos=False, device=fabric.device)
+    t = time.perf_counter() - t0
+    tokens = encoded.tolist()
+    return TokenizeResponse(tokens=tokens, request_time=t)
+
+
+@app.get("/")
```
Comment (msaroufim) [inline toy-submission/main.py]: nit: delete below this line

**Review by msaroufim — CHANGES_REQUESTED** [review]
Thanks the dataclasses in main.py made things super easy to read, left some nits - didn't actually manage to run this since I'm missing the `scripts/` directory

File: toy-submission/Dockerfile (hunk 2 above)
Comment (drisspg) [reply toy-submission/Dockerfile:21]: hmm this should be copied into from lit-llama

did you git clone recusrively?

**Comment by msaroufim** [comment]
Just wanted to confirm I tried the instructions as is and it worked

## Sample inference

Screenshot 2023-06-28 at 1 57 01 PM

## HELM inference

Also works

Screenshot 2023-06-28 at 1 58 33 PM

## Running HELM server (leaderboard)

also works

Screenshot 2023-06-28 at 2 01 05 PM

but not sure i can make sense of the metrics yet

File: toy-submission/README.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+### Make your GPUs visible to Docker 
+```
+nvidia-ctk runtime configure
+systemctl restart docker
+```
+
+### Build and run 
+```
+docker build -t toy_submission .
+docker run --gpus all -p 8080:80 toy_submission
```
Comment (msaroufim) [inline toy-submission/README.md:38]: re-constraints, we can do to run on a single GPU with only 8 CPU cores (I'd be fine with 16 or 32 too) which is what you get on AWS for an A10G

```
docker run --gpus '"device=0"' --cpus=8 -p 8080:80 toy_submission
```

Regarding disk size we can't explicitly set the contraint from running docker so instead we can add something like

```
    "storage-opts": [
      "size=500GB"
    ]
```

to ` /etc/docker/daemon.json` so the container hard errors in case it builds to something larger than 500GB

**Comment by msaroufim** [comment]
Ok this works end to end if you max out max eval instances

97% accuracy on philosophy not bad!

image

Repro instructions for those checking out this PR
* check out this branch `gh pr checkout https://github.com/stanford-crfm/helm/pull/1693`
* `cd helm` and `pip install -e .`
* Run instructions in README of this PR
* In a seperate window follow instructions from https://github.com/llm-efficiency-challenge/neurips_llm_efficiency_challenge/pull/1

```
echo 'entries: [{description: "mmlu:subject=philosophy,model=neurips/local", priority: 1}]' > run_specs.conf
helm-run --conf-paths run_specs.conf --suite v1 --max-eval-instances 1
helm-summarize --suite v1
helm-server
```

**Comment by weiweiy** [comment]
@drisspg  @msaroufim  is this ready to merge?

File: toy-submission/Dockerfile (hunk 2 above)
Comment (weiweiy) [reply toy-submission/Dockerfile:21]: added instruction in readme for git clone --recurse-submodules
